indexer/indexer.go

```diff
@@ -0,0 +1,383 @@
+	for _, v := range validatorInfos {
+		stakeCredits[v.Credit], err = stakecredit.New(v.Operator, v.Credit, client)
+		if err != nil {
+			log.Errorw("indexer: failed to new stakecredit", "err", err)
+			return nil, err
+		}
+	}
+
+	return &indexer{
+		cursor:                    cursor,
+		NumberOfBlocksForFinality: cfg.NumberOfBlocksForFinality,
+		delegators:                cfg.Delegators,
+		store:                     store,
+		client:                    client,
+		stakeHub:                  stakeHub,
+		stakeCredits:              stakeCredits,
+	}, nil
+}
+
+type indexer struct {
+	cursor                    int64
+	NumberOfBlocksForFinality int64
+	delegators                []common.Address
+	parentHeader              *types.Header
+
+	store        store.Store
+	client       *ethclient.Client
+	stakeHub     *stakehub.Contract
+	stakeCredits map[common.Address]*stakecredit.ContractWithInfo // credit address -> stake credit
+}
+
+func (i *indexer) Start(ctx context.Context) {
+	ticker := time.NewTicker(3 * time.Second)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
+			err := i.index(ctx)
+			if err != nil {
+				log.Errorw("indexer: failed to indexing", "err", err)
+			}
+		}
+	}
+}
+
+func (i *indexer) index(ctx context.Context) error {
+	var finalizedHeader *types.Header
+
+	if err := RtyDo(func() error {
+		var er error
+		finalizedHeader, er = i.client.FinalizedHeader(ctx, i.NumberOfBlocksForFinality)
+		return er
+	}); err != nil {
+		log.Errorw("indexer: failed to get finalized header", "err", err)
+		return err
+	}
+
+	for number := i.cursor + 1; number <= finalizedHeader.Number.Int64(); number++ {
+		var header *types.Header
+		if number == finalizedHeader.Number.Int64() {
+			header = finalizedHeader
+		} else {
+			if err := RtyDo(func() error {
+				var er error
+				header, er = i.client.HeaderByNumber(ctx, big.NewInt(number))
+				return er
+			}); err != nil {
+				log.Errorw("indexer: failed to get header by number", "number", number, "err", err)
+				return err
+			}
+		}
+
+		if i.parentHeader == nil {
+			if err := RtyDo(func() error {
+				var er error
+				i.parentHeader, er = i.client.HeaderByNumber(ctx, big.NewInt(number-1))
+				return er
+			}); err != nil {
+				log.Errorw("indexer: failed to get header by number", "number", number-1, "err", err)
+				return err
+			}
+		}
+
+		err := i.handleStakeHubLogs(ctx, header)
+		if err != nil {
+			log.Errorw("indexer: failed to process stakeHub logs", "err", err)
+			return err
+		}
+
+		err = i.handleStakeCreditLogs(ctx, header)
+		if err != nil {
+			log.Errorw("indexer: failed to process stakeCredit logs", "err", err)
+			return err
+		}
+
+		if isBreatheBlock(i.parentHeader.Time, header.Time) {
+			err = i.handleStakeCreditCalls(ctx, header)
+			if err != nil {
+				log.Errorw("indexer: failed to process validator and delegator", "err", err)
+				return err
+			}
+		}
+
+		err = i.store.SaveCursor(ctx, number)
```

Comment:
Because "Savexxx" of other table and `SaveCursor` is not atomic, I am afraid if panic during the time, it can not recover without any other database key conflict..

Reply:
I think it's no worry, the indexing process is idempotent